app/controllers/admin/contents_controller.rb

```diff
@@ -26,5 +20,14 @@ def update
 
       redirect_to main_app.edit_admin_content_path
     end
+
+    private
+
+    def preference_sections
+      Dir["app/models/preference_sections/*.rb"].map do |filename|
+        basename = 'PreferenceSections::' + File.basename(filename, '.rb').camelize
+        basename.constantize.new
```

Comment:
This is very clever. But I would prefer it more explicit. The problems I see:

- From the code I can't see what is loaded. I need to go to the directory to find out.
- I can't navigate to a section declaration using my IDE.
- The dependency on these classes cannot be detected automatically.
- One file not following the file name / class name convention breaks the code (class not found).
- We cannot define the order or the sections here. They may appear random depending on the file system.

Reply:
Just to give another opinion, let me answer these questions:

> From the code I can't see what is loaded. I need to go to the directory to find out.

Like anything in an Object-oriented system. And that's a great benefit; you can read what the code does at an abstract level to understand how an algorithm works. You can get into detail if you need to by reading the objects' implementations.

> I can't navigate to a section declaration using my IDE.

Indeed that is a con, but there's no direct reference to a particular section so checking the dir you'll see the section object you are interested in. No a big deal with so simple objects.

> The dependency on these classes cannot be detected automatically.

Probably the biggest downside. We just have the call to `Dir` which won't be automatically detected.

> One file not following the file name / class name convention breaks the code (class not found).

The file convention is being `*.rb` as shown in line 27. As for the class name, any name under that namespace will be picked. So unless you change this method or you don't declare the new class under the `PreferenceSections` namespace any will work.

> We cannot define the order or the sections here. They may appear random depending on the file system.

Easy to fix by using `sort_by` with a method each section object declares, and that could be even more explicit than a big array as we have now.

Reply:
>> We cannot define the order or the sections
> Easy to fix by using sort_by with a method each section object declares, and that could be even more explicit than a big array as we have now.

Adding something like priority levels to the sections so that they can be sorted creates a dependency between the sections. I can't change the position of one section without knowing the positions of all other sections. A list (array) solves this very well.

> The file convention is being *.rb

* The file needs to have the same name as the class it contains. (This fails for class names containing `API` or `URL` unless we name the files `a_p_i` and `u_r_l`.)
* They need to be in that directory. (You can't have them somewhere else and you can't have another rb file in that directory.)

These conventions are not a big deal, but I don't see any need for this flexibility here. It's a trade-off between having more code (but stupid code) and some auto-loading algorithm relying on conventions. In this case I voted for the dumb array, because it solves the sorting as well.